rl4co/models/nn/graph/gcn.py

```diff
@@ -15,7 +15,104 @@
 log = get_pylogger(__name__)
 
 
+EdgeIndexFnSignature = Callable[[TensorDict, int, bool], Tensor]
+
+
+def edge_idx_fn_wrapper(td: TensorDict, num_nodes: int):
+    # self-loop is added by GCNConv layer
+    return get_full_graph_edge_index(td.device, num_nodes, self_loop=False)
+
+
 class GCNEncoder(nn.Module):
```

Comment:
I like this clean refactoring, the logic is clearer. But will the `get_full_graph_edge_index()` be called at every forward step? i.e. in the previous version, if it's a fully connected connected graph, the `edge_index` will be saved as class variable, instead of regenerating every time.

Reply:
thats true, but the result is cached so it should not be too slow. In fact, this implementation should be much faster as before (at least it was in my experiments), because it avoids the list comprehension over the batch data within the forward pass. But I agree, its still not optimal; I will revisit this in the near future